compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -1261,6 +1255,22 @@ private void checkErrors(SymbolEnv env, BLangNode unresolvedType, BLangNode curr
         }
     }
 
+    private  boolean isTypeConstructorAvailable(NodeKind unresolvedType) {
+        switch (unresolvedType) {
+            case OBJECT_TYPE:
+            case RECORD_TYPE:
+            case CONSTRAINED_TYPE:
+            case ARRAY_TYPE:
+            case TUPLE_TYPE_NODE:
+            case TABLE_TYPE:
+            case ERROR_TYPE:
+            case FUNCTION_TYPE:
+                return true;
```

Comment:
❓ Don't we need `STREAM_TYPE`?

:wrench: The only time we call this method we don't pass `OBJECT_TYPE` is it intentional. I assume `RECORD_TYPE` should also behave the same way. Perhaps we need a different method name here and some tests for these scenarios.

Reply:
Not passing `OBJECT_TYPE` was intentional because in [[1]](https://ballerina.io/ballerina-spec/spec.html#section_5.5.3) under `5.5.3.6 Inclusion and type-ids` it says:

> It is an error for an object-type-descriptor to directly or indirectly include itself.

`RECORD_TYPE` gets defined in the very first iteration and doesn't come into the `unresolved` list and therefore do we need to restrict that as well?

Reply:
That is for object inclusion but we can have a recursive type by having the object as a field. Yes, and objects also get defined in one iteration and field are added later. This is not necessarily the first iteration.